src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -232,28 +260,35 @@ impl Stream<Push> {
         Ok(m_len + ABYTES)
     }
 
-    fn push_impl(&mut self, m: &[u8], ad: Option<&[u8]>, tag: Tag, buf: &mut Vec<u8>) -> Result<(), ()> {
+    // SAFETY: buf must have a capacity of at least m.len() + $abytes.
+    unsafe fn push_impl(&mut self, m: &[u8], ad: Option<&[u8]>, tag: Tag, buf: &mut Vec<u8>) -> Result<(), ()> {
```

Comment:
Instead of making this unsafe, why not move the two `reserve` calls into this and thereby ensure the invariant locally? (Doing `Vec::new` followed by `Vec::reserve` should not be measurably less efficient than `Vec::with_capacity`.)

Reply:
Yeah that makes sense, I'll make a commit. Thanks!